File: src/ObjectManager/Object.Tes/FilePacks/Records/045-GRAS.Grass.cs
using OA.Core;

namespace OA.Tes.FilePacks.Records
{
    public class GRASRecord : Record
    {
        public struct DATAField
        {
            public byte Density;
            public byte MinSlope;
            public byte MaxSlope;
            public ushort UnitFromWaterAmount;
            public uint UnitFromWaterType;
            //Above - At Least,
            //Above - At Most,
            //Below - At Least,
            //Below - At Most,
            //Either - At Least,
            //Either - At Most,
            //Either - At Most Above,
            //Either - At Most Below
            public float PositionRange;
            public float HeightRange;
            public float ColorRange;
            public float WavePeriod;
            public byte Flags;

            public DATAField(UnityBinaryReader r, int dataSize)
            {
                Density = r.ReadByte();
                MinSlope = r.ReadByte();
                MaxSlope = r.ReadByte();
                r.ReadByte();
                UnitFromWaterAmount = r.ReadLEUInt16();
                r.SkipBytes(2);
                UnitFromWaterType = r.ReadLEUInt32();
                PositionRange = r.ReadLESingle();
                HeightRange = r.ReadLESingle();
                ColorRange = r.ReadLESingle();
                WavePeriod = r.ReadLESingle();
                Flags = r.ReadByte();
                r.SkipBytes(3);
            }
        }

        public override string ToString() => $"GRAS: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL;
        public DATAField DATA;

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID": EDID = new STRVField(r, dataSize); return true;
                case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                case "MODB": MODL.MODBField(r, dataSize); return true;
                case "MODT": MODL.MODTField(r, dataSize); return true;
                case "DATA": DATA = new DATAField(r, dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/045-GRAS.Grass.swift
//
//  GRASRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class GRASRecord: Record, IHaveEDID, IHaveMODL {
    public struct DATAField {
        public let density: UInt8
        public let minSlope: UInt8
        public let maxSlope: UInt8
        public let unitFromWaterAmount: UInt16
        public let unitFromWaterType: UInt32
        //Above - At Least,
        //Above - At Most,
        //Below - At Least,
        //Below - At Most,
        //Either - At Least,
        //Either - At Most,
        //Either - At Most Above,
        //Either - At Most Below
        public let positionRange: Float
        public let heightRange: Float
        public let colorRange: Float
        public let wavePeriod: Float
        public let flags: UInt8

        init(_ r: BinaryReader, _ dataSize: Int) {
            density = r.readByte()
            minSlope = r.readByte()
            maxSlope = r.readByte()
            r.skipBytes(1)
            unitFromWaterAmount = r.readLEUInt16()
            r.skipBytes(2)
            unitFromWaterType = r.readLEUInt32()
            positionRange = r.readLESingle()
            heightRange = r.readLESingle()
            colorRange = r.readLESingle()
            wavePeriod = r.readLESingle()
            flags = r.readByte()
            r.skipBytes(3)
        }
    }

    public override var description: String { return "GRAS: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var MODL: MODLGroup? = nil
    public var DATA: DATAField!
    
    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID": EDID = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "MODB": MODL!.MODBField(r, dataSize)
        case "MODT": MODL!.MODTField(r, dataSize)
        case "DATA": DATA = DATAField(r, dataSize)
        default: return false
        }
        return true
    }
}
